### PR TITLE
Use Openstack os-ext-ips extension to qualify IP address types

### LINKED ISSUE #18409: kubernetes openstack cloud provider seems have hardcode network name
From https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/openstack.go 
we can find that,
```golang
func findAddrs(netblob interface{}) []string {
	// Run-time types for the win :(
	ret := []string{}
	list, ok := netblob.([]interface{})
	if !ok {
		return ret
	}
	for _, item := range list {
		props, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		tmp, ok := props["addr"]
		if !ok {
			continue
		}
		addr, ok := tmp.(string)
		if !ok {
			continue
		}
		ret = append(ret, addr)
	}
	return ret
}
```

1. the hardcode "private" and "public", seems only support that network with such name.

2. second, it fetch the first ip. usually, it can include both v6 and v4 ip. what's the meaning to just fetch the first one ? is it v4 or v6 ?

### PR DESCRIPTION
**Pull request by lebauce**
fixes #18409

### CODE REVIEW

**Comment by lebauce** [comment]
Signed the CLA

**Comment by davidopp** [comment]
@anguslees @hacktoday @ashw7n would one of you be interested in reviewing this?

**Comment by HackToday** [comment]
I may need some  time to review it, as not review kubernetes code before.  I would try to do it ASAP.  @anguslees may be more familiar with it than me.

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -352,58 +352,90 @@ func getServerByName(client *gophercloud.ServiceClient, name string) (*servers.S
...
+func getAddressesByName(client *gophercloud.ServiceClient, name string) ([]api.NodeAddress, error) {
+	srv, err := getServerByName(client, name)
+	if err != nil {
+		return nil, err
+	}
+
+	addrs := []api.NodeAddress{}
+
+	for network, netblob := range srv.Addresses {
+		var addressType api.NodeAddressType
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack.go]: Move this declaration down into the `for _, item := range list { ... }` block below where it is used - just so we can be sure we aren't mistakenly carrying over `addressType` from some earlier iteration of the below loop.

**Comment by anguslees** [comment]
One very minor code comment.  This looks good to me, thanks @lebauce!

**Comment by lebauce** [comment]
@anguslees Thanks. I pushed a new version with your suggestion.

**Comment by anguslees** [comment]
lgtm

**Comment by HackToday** [comment]
lgtm But not sure how can we test that ?  Maybe k8s include such intergation test for this change ?

**Comment by anguslees** [comment]
Theoretically, a k8s integration test could spin up a temporary openstack install (probably using devstack) easily enough.  I haven't had time to work on such a thing yet, but it would be a significant step forwards for openstack/k8s QA.  If I'm writing out my full christmas wishlist, we probably want to test against both a nova-network and a neutron -based openstack.

**Comment by lebauce** [comment]
@anguslees @HackToday I rebased my patch so this may need a new LGTM (I'm not very familiar with the pull request workflow of Kubernetes and the `Submit Queue` check seems to be stuck in pending state :-/ )

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
@justinsb can you take a quick look at this as well?

**Comment by lebauce** [comment]
@k8s-bot unit test this

**Comment by HackToday** [comment]
It looks good, my only concern is how can we easy test it to make sure it works? Is it covered by kubernetes integration test  ?

**Comment by anguslees** [comment]
I'm not familiar with debugging e2e test failures, but I gather the 3 failures in the logs above are unrelated to this change (in particular, I don't think the e2e tests actually exercise any part of the openstack plugin!).

**Comment by lebauce** [comment]
@smarterclayton As stated by @anguslees, the e2e failures seem to not be related to the PR. Is it possible to retrigger the tests ?

**Comment by smarterclayton** [comment]
ok to test

**Comment by lebauce** [comment]
Not sure who to ping here :-/ @smarterclayton ?

**Comment by smarterclayton** [comment]
@jsafrane can you review before I add the lgtm label to this?

**Comment by jsafrane** [comment]
@smarterclayton: I know nothing about OpenStack networking habbits, but the code looks ok.

**Comment by lebauce** [comment]
@smarterclayton Thanks.

Could you please ask to run the tests again please ?

**Comment by smarterclayton** [comment]
@k8s-bot unit test this

**Comment by akram** [comment]
Hi,
I am interesting in getting this merged as well.
we do have a reproducible issue.

**Comment by ncdc** [comment]
@k8s-bot test this

**Pull request merged**